articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -0,0 +1,188 @@
+It will receive as a parameter a `TopicStatusEvent` enum value that specifies what type of event took place and a timestamp of when the event took place.
+This callback function would be provided by the user's application when it calls the create function for publishers and subscribers.
+The constructors and create functions will be overloaded to make this new handler optional.
+
+The choice to use a callback function as opposed to another notification mechanism is to remain consistent with the existing ROS interface and not add a new method of waiting on ROS data/events.
+
+#### QoS Struct
+
+Minimal changes will need to be made to the QoS struct that is passed into the creation functions for Topics, Services, and Actions.
+A couple new enum values will be added for the Deadline and Liveliness settings and then a couple integers will be added to specify the time values for these policies.
+
+#### Assert Liveliness Functions
+
+Two new functions would need to be added that could be used by the application to explicitly assert liveliness.
+One function to assert liveliness at the Node level and one to assert it at the Topic level.
+While liveliness will also be implicitly assumed just based on sending messages, it will also be able to be explicitly declared by the application.
+These functions will need to be implemented in the rmw layer, the rcl layer, and the language specific client libraries.
+
+#### rcl_wait and rmw_wait
+
+The rcl layer is currently using a WaitSet in order to be informed of events from the rmw layer, such as incoming messages.
+These WaitSets contain lists of several types of conditions, such as timers firing or subscriptions receiving data.
+In order to support new Topic Status Events, a new type will be added to the existing WaitSet and the rmw layer will set them when these events occur.
+
+
+## FAQ
+
+- How does the Deadline policy take into account the additional overhead of ROS (such as deserialization) when determining if a deadline was missed?
+  - As a simplification it is not going to attempt to take into account any ROS overhead. A deadline will be considered missed if the rmw layer does not receive a message by the deadline and not if the user application on top of ROS does not receive it by the deadline.
+
+
+## Open Questions
+
+- How do Deadlines and Liveliness impact Services and Actions?
+  - Actions and Services are slightly different in that they are composed of multiple Topics. It may make sense for a Service to specify a Liveliness policy on its own DataWriter, but it likely doesn't need the same policy on the DataReader.
+  - One option would be to only apply the QoS settings to one or another of the topics. So Liveliness may only apply to the DataWriters in Services and Actions and the Deadlines might apply only to the "update" topic data writer for Actions.
```

Comment:
Robustness needs mean that all topics involved should have liveliness checks. If something bizarre happens and just one topic goes away, we need to know.

Reply:
In some cases I agree, but I think there are other cases where your client cares if the service is no longer alive, but the service likely doesn't care that clients are no longer alive in most situations. 

In the next iteration I'm including more details about the policy settings and how it impacts topics and services separately. I'm working under the assumption that services are not necessarily two topics either, since that seems to be dependent on the rmw implementation at the moment (even though some recommendations are given to rmw layers as to how to implement it to be compatible). Correct me if I'm wrong about that assumption.

Reply:
I think assuming that services are not necessarily two topics is a solid assumption. If someone were deranged enough to implement an RMW for CORBA, for example...

However this makes it more necessary, not less, that the requirements placed on the RMW allow that all topics, services, RPC mechanisms or whatever are guaranteed to be alive.

"Know that all parts are alive" is a valid use case. If "the service not caring that clients are no longer alive but clients caring" is a valid use case, too, then this shows that we need to be able to specify QoS settings for each side differently.

Reply:
I've made a recommendation in the design to start using a separate struct to define the QoS policies for Services and Topics. I also recommended adding additional settings into the QoS policy for Services to define the behavior desired from the client and the behavior desired by the owner. That should provide the flexibility for the applications to do what they want.